lib/linked_list.rb

```diff
@@ -12,152 +12,307 @@ def initialize(value, next_node = nil)
-    def find_nth_from_end(n)
-      raise NotImplementedError
+    return count_length + 1
+  end
+
+  # method that returns the value at a given index in the linked list
+  # index count starts at 0
+  # returns nil if there are fewer nodes in the linked list than the index value
+  # Time Complexity: O(n), n is the length of the list
+  # Space Complexity: O(1)
+  def get_at_index(index)
+    return nil if @head == nil
+
+    current = @head
+    counter = 0
+
+    until counter == index
+      if current.next == nil
+        return nil
+      else
+        current = current.next
+        counter += 1
+      end
     end
+    return current.data
+  end
+
+  # method to print all the values in the linked list
+  # Time Complexity: O(n), n is the number of nodes in the LL
+  # Space Complexity: O(1)
+  def visit
+    current = @head
 
-    # checks if the linked list has a cycle. A cycle exists if any node in the
-    # linked list links to a node already visited.
-    # returns true if a cycle is found, false otherwise.
-    # Time Complexity:  
-    # Space Complexity
-    def has_cycle
-      raise NotImplementedError
+    until current.next == nil
+      puts current.data
+      current = current.next
     end
+    puts current.data
+  end
+
+  # method to delete the first node found with specified value
+  # Time Complexity: O(n), n is the number of nodes in the LL
+  # Space Complexity: O(1)
+  def delete(value)
+    return if @head == nil
 
+    previous_node = nil
+    current = @head
+    next_node = nil
 
-    # Additional Exercises 
-    # returns the value in the first node
-    # returns nil if the list is empty
-    # Time Complexity:  
-    # Space Complexity
-    def get_first
-      raise NotImplementedError
+    if current.data == value
+      @head = current.next
+      current.next = nil
+      return
     end
 
-    # method that inserts a given value as a new last node in the linked list
-    # Time Complexity:  
-    # Space Complexity
-    def add_last(value)
-      raise NotImplementedError
+    while current.data != nil
+      if current.next.data == value
+        node_to_delete = current.next
+        current.next = node_to_delete.next
+        node_to_delete.next = nil
+        return
+      else
+        current = current.next
+      end
     end
+  end
+
+  # method to reverse the singly linked list
+  # note: the nodes should be moved and not just the values in the nodes
+  # Time Complexity: O(n), n is the number of nodes in the LL
+  # Space Complexity: O(1)
+  def reverse
+    current = @head
+    temp_a = current
+    temp_b = current.next
+    current.next = nil
+    current = temp_b
 
-    # method that returns the value of the last node in the linked list
-    # returns nil if the linked list is empty
-    # Time Complexity:  
-    # Space Complexity
-    def get_last
-      raise NotImplementedError
+    while current.next != nil
+      temp_b = current.next
+      current.next = temp_a

```

Comment:
Wait, `temp_a` starts as `current`, so you're going to end up making `current` point to itself.  It think it would be better to start (initialize) `temp_a` to `nil`.

I also suggest renaming the temps.  Maybe make one `next_node` and rename `temp_a` to `previous_node`